### DISCUSSION TITLE
"This session already exists." error with cy.session() command in test runner.

### DISCUSSION THREAD

**Discussion by steperry** [cypress-io/cypress#30674] [Questions and Help]
Hi, I've been using the cy.session() command recently and while it seems useful, I've run into some simple issues that will likely force me to abandon it. Anyone know if these are bugs, or am I just not using it correctly?

Sometimes this error will appear in the test runner: "This session already exists. You may not create a new session with a previously used identifier."

I understood that the command would automatically restore the existing session with the same name, so what would cause it to return an error like this? I am using the cacheAcrossSpecs option because all of my tests should be able to share the same session. If cypress has determined the session is no longer valid, should it not just recreate it?

Additionally, when this error happens in the test runner, there is no longer a "Clear All Sessions" button, and this is probably the time when it would be most useful. If I change my session name so the test starts working again, I regain access to the "Clear All Sessions" button.

Thanks for any help.

**Comment by portyu9** [comment]
Hi @steperry 

A failed `validate()` is **not** what causes this error. If a cached session fails validation, Cypress should rerun `setup()` and recreate it.

`This session already exists... previously used identifier` means Cypress has already seen that session ID with a **different session definition**. With `cacheAcrossSpecs: true`, every spec that uses the cached session must use the same `id`, `setup`, `validate`, and `cacheAcrossSpecs` value.

The safest pattern is to define `cy.session()` once in a shared login command/helper and call that helper from every spec instead of copying the session definition around.

Also give genuinely different login flows different IDs, for example:

```js
cy.session(['loginByApi', username], setupApiLogin, options)
cy.session(['loginByForm', username], setupFormLogin, options)
```

For debugging in open mode you can clear the backend cache programmatically:

```js
Cypress.session.clearAllSavedSessions()
```

So I would look for another `cy.session()` call using the same ID with a different callback/options, or for a session definition that was edited while the open-mode cache was still alive.

Please mark this answer as accepted if it helped, thank you.
